### PR TITLE
[AWS] Add API Gateway metrics data stream 

### LINKED ISSUE #6270: [AWS API Gateway] Metrics datastream
## Summary

Collect metrics, analyze them and produce an integration for API Gateway metrics. Analyzing these metrics helps detect bottlenecks, optimize resources, and troubleshoot issues.

This integration will include all three supported API types: HTTP, REST, and WebSocket. 

## Tasks

- [x] Set up or use existing instances for the three API types to generate metrics
- [x] Initially inspect metrics for confirmation via [AWS CloudWatch](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/WhatIsCloudWatch.html) or [AWS CLI](https://aws.amazon.com/cli/)
- [x] Fetch metrics via [metricbeat](https://www.elastic.co/beats/metricbeat) and save sample event
- [x] Create the integration in accordance to fetched data which includes fields, dimensions, sample event, icon etc.
- [x] Spin up the stack locally; Set up and verify the integration via Kibana
- [ ] Create documentation which includes summary, setup, requirements, fields and sample event

## Challenges

- AWS naming scheme can be confusing (ex: `4XXError` for REST vs `4xx` for HTTP). Both metrics exist and appear to serve the exact same purpose but it's easy to assume one is just a typo initially.
- Given that no metrics are currently being generated in the Elastic account for API Gateway, metricbeat might not fetch metrics consistently. This is because they miss the [time window](https://github.com/elastic/beats/blob/f65554fdfb439f71f7107524c4c050767f2e9bb7/x-pack/metricbeat/module/aws/utils.go#L32-L37) that the CloudWatch metricset uses.

## Docs

Metrics and dimensions documentation:
- [HTTP API](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-metrics-and-dimensions.html) 
- [REST API](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-metrics.html) 
- [WebSocket](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-websocket-api-logging.html)

### PR DESCRIPTION
**Pull request by lucianpy**
## What does this PR do?
Adds API Gateway metrics datastream which covers HTTP, REST and Websocket APIs.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author checklist
- [x] Add unit metric types
- [x] Add docs

## Related issues

Concerns https://github.com/elastic/integrations/issues/6058
Closes #6270

### CODE REVIEW

File: packages/aws/data_stream/apigateway_metrics/agent/stream/stream.yml.hbs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+{{#if tags_filter}}
+tags_filter: {{tags_filter}}
+{{/if}}
+{{#if proxy_url }}
+proxy_url: {{proxy_url}}
+{{/if}}
+metrics:
+- namespace: AWS/ApiGateway
+  statistic: ["Sum"]
+  name:
```
Comment (shmsr) [inline packages/aws/data_stream/apigateway_metrics/agent/stream/stream.yml.hbs:45]: If I look at this doc: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-metrics-and-dimensions.html#:~:text=CloudWatch%20metrics.-,API%20Gateway%20metrics,-Amazon%20API%20Gateway

I do not see `5xx`, `4xx`, etc. Also in the doc, I see `Latency` which is not here. Is there documentation not up to date?

File: packages/aws/data_stream/apigateway_metrics/fields/fields.yml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+              description: The amount of data processed in bytes.
+            - name: CacheHitCount.sum
+              type: long
+              description: The number of requests served from the API cache in a given period.
+            - name: CacheMissCount.sum
+              type: long
+              description: The number of requests served from the backend in a given period, when API caching is enabled.
+            - name: Count.sum
+              type: long
+              description: The total number API requests in a given period.
```
Comment (shmsr) [inline packages/aws/data_stream/apigateway_metrics/fields/fields.yml:42]:
```suggestion
              description: The total number of API requests in a given period.
```

File: packages/aws/data_stream/apigateway_metrics/fields/fields.yml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+              description: The number of requests that have a 4XX response returned by API Gateway before the integration is invoked.
+            - name: ExecutionError.sum
+              type: long
+              description: Errors that occurred when calling the integration.
+    - name: dimensions
+      type: group
+      fields:
+        - name: ApiId
+          type: keyword
+          description: Each API created in API Gateway is assigned a unique ApiId, which is used to distinguish and reference that specific API within the system.
```
Comment (shmsr) [inline packages/aws/data_stream/apigateway_metrics/fields/fields.yml:63]: In the doc, I also don't see this. Can you please link from did you get this dimension?

File: packages/aws/data_stream/apigateway_metrics/fields/package-fields.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+- name: aws
+  type: group
+  fields:
+    - name: tags.*
+      type: object
+      description: |
+        Tag key value pairs from aws resources.
```
Comment (shmsr) [inline packages/aws/data_stream/apigateway_metrics/fields/package-fields.yml:7]:
```suggestion
        Tag key-value pairs from AWS resources.
```

File: packages/aws/data_stream/apigateway_metrics/sample_event.json [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                },
+                "Count": {
+                    "sum": 2
+                },
+                "IntegrationLatency": {
+                    "avg": 85.5
+                },
+                "Latency": {
+                    "avg": 104
+                }
```
Comment (shmsr) [inline packages/aws/data_stream/apigateway_metrics/sample_event.json:79]: See `Latency` is here?

File: packages/aws/data_stream/apigateway_metrics/sample_event.json [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    },
+    "aws": {
+        "apigateway": {
+            "metrics": {
+                "4xx": {
+                    "sum": 0
+                },
+                "5xx": {
+                    "sum": 0
+                },
```
Comment (shmsr) [inline packages/aws/data_stream/apigateway_metrics/sample_event.json:67]: Ah, here I see `4xx` and `5xx`.

File: packages/aws/manifest.yml [hunk 7]
```diff
@@ -739,5 +739,22 @@ policy_templates:
         title: Amazon GuardDuty logo
         size: 33x39
         type: image/svg+xml
+  - name: apigateway
+    title: Amazon API Gateway
+    description: Collect logs and metrics for Amazon API Gateway with Elastic Agent
```
Comment (shmsr) [inline packages/aws/manifest.yml:744]:
```suggestion
    description: Collect logs and metrics for Amazon API Gateway with Elastic Agent.
```

File: packages/aws/manifest.yml [hunk 8]
```diff
@@ -739,5 +739,22 @@ policy_templates:
...
+    data_streams:
+      - apigateway_metrics
+    categories:
+      - observability
+    inputs:
+      - type: aws/metrics
+        title: Collect API Gateway metrics
+        description: Collect API Gateway metrics using AWS CloudWatch
```
Comment (shmsr) [inline packages/aws/manifest.yml:752]:
```suggestion
        description: Collect API Gateway metrics using AWS CloudWatch.
```

File: packages/aws/data_stream/apigateway_metrics/sample_event.json (hunk 6 above)
Comment (shmsr) [reply packages/aws/data_stream/apigateway_metrics/sample_event.json:67]: Also, saw this: https://github.com/aws/aws-cdk/blob/b885ece7285ebe8219dbe4d62ae80554bd657632/packages/%40aws-cdk/aws-apigatewayv2-alpha/lib/http/stage.ts#L107

Comment (shmsr) [reply packages/aws/data_stream/apigateway_metrics/sample_event.json:67]: Maybe there's no `4xxError` or `5xxError` metric. Maybe a mistake in their doc?

**Review by shmsr — CHANGES_REQUESTED** [review]
Left some comments.

**Comment by shmsr** [comment]
> Member

Oh, okay. Thanks for sharing!

File: packages/aws/data_stream/apigateway_metrics/fields/package-fields.yml (hunk 4 above)
Comment (lucianpy) [reply packages/aws/data_stream/apigateway_metrics/fields/package-fields.yml:7]: This one is always the same, if you want, we can improve it starting with this PR.

File: packages/aws/docs/apigateway.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
+# AWS API Gateway
+
+The AWS API Gateway integration allows you to monitor [API Gateway](https://aws.amazon.com/api-gateway/)—a centralized API management service.
```
Comment (shmsr) [inline packages/aws/docs/apigateway.md:3]:
```suggestion
The AWS API Gateway integration allows you to monitor [API Gateway](https://aws.amazon.com/api-gateway/) — a centralized API management service.
```

File: packages/aws/_dev/build/docs/apigateway.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
+# AWS API Gateway
+
+The AWS API Gateway integration allows you to monitor [API Gateway](https://aws.amazon.com/api-gateway/)—a centralized API management service.
```
Comment (shmsr) [inline packages/aws/_dev/build/docs/apigateway.md:3]:
```suggestion
The AWS API Gateway integration allows you to monitor [API Gateway](https://aws.amazon.com/api-gateway/) — a centralized API management service.
```

**Review by shmsr — CHANGES_REQUESTED** [review]
Looks good to me. 

But as I'm still not much experienced with integrations as of now, I'd like to request you get another pair of eyes to review this PR. I might have missed something.

**Comment by gpop63** [comment]
LGTM, I'll take a look at the documentation.

File: packages/aws/data_stream/apigateway_metrics/fields/fields.yml [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        - name: metrics
+          type: group
+          fields:
+            - name: 4XXError.sum
+              type: long
+              description: The number of client-side errors captured in a given period.
+              metric_type: counter
+            - name: 4xx.sum
+              type: long
+              description: The number of client-side errors captured in a given period.
```
Comment (zmoog) [inline packages/aws/data_stream/apigateway_metrics/fields/fields.yml:16]: To check my understanding after reading the comments: `4XXError` and `4xx` have the same semantics, but they come from different API types (REST vs. HTTP) right?

File: packages/aws/data_stream/apigateway_metrics/manifest.yml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+        multi: false
+        required: false
+        show_user: false
+      - name: regions
+        type: text
+        title: Regions
+        multi: true
+        required: false
+        show_user: true
+      - name: latency
```
Comment (zmoog) [inline packages/aws/data_stream/apigateway_metrics/manifest.yml:25]: Stealing the `latency` description from https://github.com/elastic/integrations/pull/6231 

```suggestion
      - name: latency
        description: To address latency issues between certain AWS services and CloudWatch, specify a latency parameter to adjust the collection start time and end time in Metricbeat such as `5m`.
```

**Review by zmoog — APPROVED** [review]
Please considers adding the latency doc from the other PR. 

LGTM.

File: packages/aws/data_stream/apigateway_metrics/fields/fields.yml (hunk 11 above)
Comment (lucianpy) [reply packages/aws/data_stream/apigateway_metrics/fields/fields.yml:16]: Correct @zmoog

**Review by gpop63 — APPROVED** [review]

**Review by shmsr — APPROVED** [review]

**Pull request merged**